### PR TITLE
Add native stringview support for RIGHT

### LINKED ISSUE #11917: Update `RIGHT` scalar function to support Utf8View
Part of https://github.com/apache/datafusion/issues/11752 and https://github.com/apache/datafusion/issues/11790

Currently, a call to `RIGHT` with a Utf8View datatypes induces a cast. After the change that fixes this issue, it should not.

```sql
query TT
EXPLAIN SELECT
  RIGHT(column1_utf8view, 3) as c2
FROM test;
----
logical_plan
01)Projection: right(CAST(test.column1_utf8view AS Utf8), Int64(3)) AS c2
02)--TableScan: test projection=[column1_utf8view]
```

`right` is defined here: https://github.com/apache/datafusion/blob/e088945c38b74bb1d86dcbb88a69dfc21d59e375/datafusion/functions/src/unicode/right.rs

casting tests are in: https://github.com/apache/datafusion/blob/main/datafusion/sqllogictest/test_files/string_view.slt

### Is your feature request related to a problem or challenge?

We are working to add complete StringView support in DataFusion, which permits potentially much faster processing of string data. See https://github.com/apache/datafusion/issues/10918 for more background.

Today, most DataFusion string functions support DataType::Utf8 and DataType::LargeUtf8 and when called with a StringView argument DataFusion will cast the argument back to DataType::Utf8 which is expensive.

To realize the full speed of StringView, we need to ensure that all string functions support the DataType::Utf8View directly.

### Describe the solution you'd like

Update the function to support DataType::Utf8View directly

### Describe alternatives you've considered

The typical steps are:
1. Write some tests showing the function doesn't support Utf8View (see the tests in [`string_view.slt`](https://github.com/apache/datafusion/blob/main/datafusion/sqllogictest/test_files/string_view.slt) to ensure the arguments are not being cast
2. Change the `Signature` of the function to accept `Utf8View` in addition to `Utf8`/`LargeUtf8`
3. Update the implementation of the function to operate on `Utf8View`

Example PRs
* Update to use an arrow kernel that already supports StringView:  https://github.com/apache/datafusion/pull/11787
* Change the implementation to support StringView directly: https://github.com/apache/datafusion/pull/11676
* Change implementation (option 2): https://github.com/apache/datafusion/pull/11556

### Additional context

The documentation of string functions can be found here: https://datafusion.apache.org/user-guide/sql/scalar_functions.html#string-functions

To test a function with StringView with `datafusion-cli` you can use an example like this (replacing `starts_with` with the relevant function)
```sql
> create table foo as values (arrow_cast('foo', 'Utf8View'), arrow_cast('bar', 'Utf8View'));
0 row(s) fetched.
Elapsed 0.043 seconds.

> select starts_with(column1, column2) from foo;
+--------------------------------------+
| starts_with(foo.column1,foo.column2) |
+--------------------------------------+
| false                                |
+--------------------------------------+
1 row(s) fetched.
Elapsed 0.015 seconds.
```

To see if it is using utf8 view, use `EXPLAIN` to see the plan and verify there is no `CAST`. In this example the `CAST(column1@0 AS Utf8)` indicates that the function is not using `Utf8View` natively

```sql
> explain select starts_with(column1, column2) from foo;
+---------------+------------------------------------------------------------------------------------------------------------------------------+
| plan_type     | plan                                                                                                                         |
+---------------+------------------------------------------------------------------------------------------------------------------------------+
| logical_plan  | Projection: starts_with(CAST(foo.column1 AS Utf8), CAST(foo.column2 AS Utf8))                                                |
|               |   TableScan: foo projection=[column1, column2]                                                                               |
| physical_plan | ProjectionExec: expr=[starts_with(CAST(column1@0 AS Utf8), CAST(column2@1 AS Utf8)) as starts_with(foo.column1,foo.column2)] |
|               |   MemoryExec: partitions=1, partition_sizes=[1]                                                                              |
|               |                                                                                                                              |
+---------------+------------------------------------------------------------------------------------------------------------------------------+
2 row(s) fetched.
Elapsed 0.006 seconds.
```

It is also often good to test with a constant as well (likewise there should be no cast):

```sql
> explain select starts_with(column1, 'foo') from foo;
+---------------+----------------------------------------------------------------------------------------------------------+
| plan_type     | plan                                                                                                     |
+---------------+----------------------------------------------------------------------------------------------------------+
| logical_plan  | Projection: starts_with(CAST(foo.column1 AS Utf8), Utf8("foo"))                                          |
|               |   TableScan: foo projection=[column1]                                                                    |
| physical_plan | ProjectionExec: expr=[starts_with(CAST(column1@0 AS Utf8), foo) as starts_with(foo.column1,Utf8("foo"))] |
|               |   MemoryExec: partitions=1, partition_sizes=[1]                                                          |
|               |                                                                                                          |
+---------------+----------------------------------------------------------------------------------------------------------+
2 row(s) fetched.
Elapsed 0.002 seconds.

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by Kev1n8**
## Which issue does this PR close?

Closes #11917

## What changes are included in this PR?
Update signature of RIGHT to accept Utf8View and add some tests for that.

## Are these changes tested?
yes

## Are there any user-facing changes?
no

### CODE REVIEW

**Comment by Kev1n8** [comment]
Pasting a possible implementation of `StringViewArray` output here (current implementation returns `StringArray`):
```
fn string_view_right(args: &[ArrayRef]) -> Result<ArrayRef> {
    let string_array = as_string_view_array(&args[0])?;
    let n_array = as_int64_array(&args[1])?;

    let mut builder = StringViewBuilder::new();
    let combined = string_array.iter().zip(n_array);

    for iter in combined {
        if let (Some(str), Some(n)) = iter {
            match n.cmp(&0) {
                Ordering::Less => {
                    let start = n.unsigned_abs() as usize;
                    let sub_string = &str[start..];
                    builder.append_value(sub_string);
                }
                Ordering::Equal => {
                    builder.append_null();
                }
                Ordering::Greater => {
                    let start = max(str.chars().count() as i64 - n, 0) as usize;
                    let sub_string = &str[start..];
                    builder.append_value(sub_string);
                }
            }
        } else {
            builder.append_null();
        }
    }
    let result = builder.finish();
    Ok(Arc::new(result) as ArrayRef)
}
```

File: datafusion/functions/src/unicode/right.rs [hunk 1]
```diff
@@ -83,6 +93,47 @@ impl ScalarUDFImpl for RightFunc {
...
+}
+
+// Currently the return type can only be Utf8 or LargeUtf8, to reach fully support, we need
+// to edit the `get_optimal_return_type` in utils.rs to make the udfs be able to return Utf8View
+// See https://github.com/apache/datafusion/issues/11790#issuecomment-2283777166
+fn string_view_right(args: &[ArrayRef]) -> Result<ArrayRef> {
+    let string_array = as_string_view_array(&args[0])?;
+    let n_array = as_int64_array(&args[1])?;
+
+    let result = string_array
```
Comment (XiangpengHao) [inline datafusion/functions/src/unicode/right.rs:110]: Code below is quite similar to the `string_right` function, I wonder if it is possible to use the `ArrayAccessor` to reuse the code:
https://github.com/apache/arrow-rs/blob/a693f0f9c37567b2b121e261fc0a4587776d5ca4/arrow-array/src/array/mod.rs#L438-L531

Similar to this pr: https://github.com/apache/datafusion/pull/11884/files#diff-e49633ca67d4035f244e96909aa4d9d9a484a933a7783bcdee3a6cf32dcd7ab8R76

Comment (Kev1n8) [reply datafusion/functions/src/unicode/right.rs]: Thanks for the advice. I couldn't come up with a proper way to combine them but now I get it :)

File: datafusion/sqllogictest/test_files/string_view.slt [hunk 2]
```diff
@@ -809,16 +809,28 @@ logical_plan
...
 EXPLAIN SELECT
   RIGHT(column1_utf8view, 3) as c2
 FROM test;
 ----
 logical_plan
-01)Projection: right(CAST(test.column1_utf8view AS Utf8), Int64(3)) AS c2
+01)Projection: right(test.column1_utf8view, Int64(3)) AS c2
 02)--TableScan: test projection=[column1_utf8view]
 
+# Test outputs of RIGHT
```
Comment (alamb) [inline datafusion/sqllogictest/test_files/string_view.slt:821]: 👍

**Review by alamb — APPROVED** [review]
Thank you @Kev1n8  and @XiangpengHao  -- very nice

**Pull request merged**